File: src/ts/src/ads/internal/AsyncHelper.ts
import { IAsyncHelper } from "./IAsyncHelper";

export class AsyncHelper<Result> implements IAsyncHelper<Result> {
    private _promise?: Promise<Result>;
    private _resolver?: (result: Result) => void;
    private _finalizer?: (result: Result) => void;
    private _processing: boolean;

    public constructor() {
        this._processing = false;
    }

    public get isProcessing(): boolean {
        return this._processing;
    }

    public process(processor: () => void, finalizer: (result: Result) => void): Promise<Result> {
        if (this._processing) {
            // Waiting.
        } else {
            this._promise = new Promise<Result>(resolver => {
                this._processing = true;
                this._finalizer = finalizer;
                processor();
            });
        }
        return this._promise!;
    }

    public resolve(result: Result): void {
        this._finalizer && this._finalizer(result);
        this._finalizer = undefined;
        this._resolver && this._resolver(result);
        this._resolver = undefined;
        this._processing = false;
    }

}

File: src/unity/Runtime/Ads/Internal/AsyncHelper.cs
using System;
using System.Threading.Tasks;

namespace EE.Internal {
    internal class AsyncHelper<Result> : IAsyncHelper<Result> {
        private TaskCompletionSource<Result> _source;
        private Action<Result> _finalizer;

        public bool IsProcessing { get; private set; }

        public Task<Result> Process(Action processor, Action<Result> finalizer) {
            if (IsProcessing) {
                // Waiting.
            } else {
                _source = new TaskCompletionSource<Result>();
                IsProcessing = true;
                _finalizer = finalizer;
                processor();
            }
            return _source.Task;
        }

        public void Resolve(Result result) {
            // Keep copies.
            var finalizer = _finalizer;
            var source = _source;

            // Update state.
            _finalizer = null;
            _source = null;
            IsProcessing = false;

            // Process.
            finalizer(result);
            source.SetResult(result);
        }
    }
}